### PR TITLE
Added the fundamental spaces

### PR DESCRIPTION
**Pull request by kraftpunk97-zz**
PR with reference to this issue #5 
seed and copy functions still need to be implemented. Modified CartPole.jl (CartPole-v1.1.jl) to run on these spaces.

### CODE REVIEW

File: src/Spaces/discrete.jl [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+self.observation_space = Discrete(2)
+"""
+mutable struct Discrete <: AbstractSpace
+    n::Int
+    dtype::DataType
+
+    Discrete(N::Int) = new(N, Int64)
+end
+
+sample(self::Discrete) = rand(1:self.n)
```
Comment (tejank10) [inline src/Spaces/discrete.jl]: `self` -> `d` (or anything else which would imply that given object is a `Discrete` when read)

File: src/Spaces/discrete.jl [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+function contains(self::Discrete, x::Union{Number, Array})
+    as_int::Union{Number, Array, Nothing} = nothing
+    try
+        as_int = Int.(x)
+    catch InexactError
+        return false
+    end
```
Comment (tejank10) [inline src/Spaces/discrete.jl:23]: Clever use of try catch!

File: src/Spaces/multi-binary.jl [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+mutable struct MultiBinary <: AbstractSpace
+    n::Int
+    dtype::DataType
+    shape::Tuple
+    MultiBinary(n::Int) = new(n, Int, (n, ))
+end
+
+sample(self::MultiBinary) = (self.dtype).(rand(0:1, self.n))
+
+contains(self::MultiBinary, x) = all(boolor.(x .== 0, x .== 1))
```
Comment (tejank10) [inline src/Spaces/multi-binary.jl]: Any particular reason for introducing `boolor`? Can we do `all(x .== 0 .| x .== 1)`?

File: src/Spaces/multi-binary.jl [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+Base.:(==)(self::MultiBinary, other::MultiBinary) = return self.n == other.n
```
Comment (tejank10) [inline src/Spaces/multi-binary.jl]: `return` can be dropped here

**Review by tejank10 — CHANGES_REQUESTED** [review]
Thanks @kraftpunk97 for great work!

File: src/Spaces/discrete.jl (hunk 1 above)
Comment (kraftpunk97-zz) [reply src/Spaces/discrete.jl]: @tejank10  A little confused here. Are you suggesting something like `sample(discrete_obj::Discrete) = rand(1:discrete_obj.n)`?

File: src/Spaces/discrete.jl (hunk 2 above)
Comment (kraftpunk97-zz) [reply src/Spaces/discrete.jl:23]: 👌👍

File: src/Spaces/multi-binary.jl (hunk 3 above)
Comment (kraftpunk97-zz) [reply src/Spaces/multi-binary.jl]: Sorry about that, I was just testing something and I forgot to remove that function. Consider it fixed...

**Comment by kraftpunk97-zz** [comment]
@tejank10 Made changes as requested. Please review and verify.

File: src/Spaces/discrete.jl (hunk 1 above)
Comment (tejank10) [reply src/Spaces/discrete.jl]: Yeah, something like that.

**Review by tejank10 — APPROVED** [review]

**Comment by kraftpunk97** [comment]
If the changes are approved then please merge. Thanks.

**Comment by tejank10** [comment]
Great! I intend to replace CartPole with CartPole1.1, if the examples in [model-zoo](https://github.com/FluxML/model-zoo/tree/master/games/differentiable-programming/cartpole) are not breaking.

**Comment by kraftpunk97-zz** [comment]
Made the necessary changes. Please verify.

**Comment by tejank10** [comment]
Great work @kraftpunk97 , Thanks!

**Comment by kraftpunk97-zz** [comment]
I have never experienced joy like this before. Thanks for sticking it out with me.

**Pull request merged**